lib/pbench/server/auth/__init__.py

```diff
@@ -0,0 +1,435 @@
+            }
+        """
+
+        if token:
+            self.add_header_param("Authorization", f"Bearer {token}")
+
+        return self._get(self.USERINFO_ENDPOINT).json()
+
+    def logout(self, refresh_token: str) -> HTTPStatus:
+        """
+        The logout endpoint logs out the authenticated user.
+        Args:
+            refresh_token: Refresh token issued at the time of login
+        Returns:
+            Https status code based on success or failure
+        """
+        if self.LOGOUT_ENDPOINT:
+            payload = {
+                "client_id": self.client_id,
+                "client_secret": self.client_secret_key,
+                "refresh_token": refresh_token,
+            }
+
+            return HTTPStatus(
+                self._post(self.LOGOUT_ENDPOINT, data=payload).status_code
+            )
+        else:
+            self.logger.warning("logout attempt on third party token")
+            abort(
+                HTTPStatus.METHOD_NOT_ALLOWED,
+                message="Logout operation is not allowed for the given token",
+            )
+
+    def revoke_access_token(self, access_token: str) -> HTTPStatus:
+        """
+        Revoke endpoint to revoke the current access token. It does not however,
+        logs the refresh token out.
+        Args:
+            access_token: token to revoke
+        Returns:
+            Https status code based on success or failure
+        """
+        payload = {
+            "client_id": self.client_id,
+            "client_secret": self.client_secret_key,
+            "token": access_token,
+            "token_type_hint": "access_token",
+        }
+
+        return HTTPStatus(
+            self._post(self.REVOCATION_ENDPOINT, data=payload).status_code
+        )
+
+    def _get(self, path: str, **kwargs) -> requests.Response:
+        """Submit get request to the path.
+        Args:
+            path: Path for the request.
+        Returns:
+            Response from the request.
+        """
+
+        try:
+            return self.connection.get(
+                urljoin(self.server_url, path),
+                params=kwargs,
+                headers=self.headers,
+                verify=self.verify,
+            )
+        except Exception as exc:
+            self.logger.exception("{}", str(exc))
+            raise OidcError(
+                HTTPStatus.INTERNAL_SERVER_ERROR, f"Can't connect to server {exc}"
+            )
+
+    def _post(self, path: str, data: Dict, **kwargs) -> requests.Response:
+        """Submit post request to the path.
+        Args:
+            path: Path for the request.
+        Returns:
+            Response from the request.
+        """
+        try:
+            return self.connection.post(
+                urljoin(self.server_url, path),
+                params=kwargs,
+                data=data,
+                headers=self.headers,
+                verify=self.verify,
+            )
+        except Exception as exc:
```

Comment:
> It seems odd to me that we don't check the response returned by the HTTP method call. Can't it fail? I don't see any setting which causes failures to be reported as exceptions...did I miss it?

What do you mean by failure? Do you mean the response returned by the HTTP method other than 200? In any case, the returned response payload is always a dictionary with a message and an appropriate status code. I was thinking to let the top-level caller (We don't have those yet) of these functions decide what to do with the response payload and status code. Also, take a look at this comment [comment](https://github.com/distributed-system-analysis/pbench/pull/2942#discussion_r945424775)

Reply:
> In any case, the returned response payload is always a dictionary with a message and an appropriate status code.

That is exactly the presumption that I'm challenging:
- At the lowest level, the code is calling `requests.Session.post()` or a similar HTTP method function; I assume that the `requests` function can fail in certain ways, some of which are reported as exceptions and some of which are reported via the `Response` object which they return.
- Assuming that the request is successfully _issued_, the resulting HTTP response may indicate that the request failed.
- The code one level up which calls your `_post()` (or equivalent) wrapper functions expects to get a `Response` object, on which it immediately invokes the `json()` method; however, I'm not convinced that the response will inevitably contain a JSON payload (especially if the request failed).
- And, finally, it's possible that all of the above works, but the API protocol provides the failure indication in the JSON payload.

So, I think we need to have in place appropriate handling for errors at each of those levels.

- I believe that failure at the first level indicates a bug in our code or a problem in the deployment configuration, but those are nevertheless real possibilities.  It's true that Flask provides the ultimate fallback for handling exceptions, but we probably want to do better than that (i.e., the default `500`/`Internal Error` is pretty useless in terms of diagnosing the situation).
- Whether we want the `_post()` (et al.) wrapper function(s) to handle HTTP protocol-level issues is an open question.  In the general case, I would probably say "no", but I don't think these wrappers are really intended for general use, and so they can afford to be opinionated.  For instance, I would expect that they should see `200` (and maybe `201`) responses, and anything else constitutes a problem of some sort.  (I assume that the `Requests` package should be told to handle `3XX` responses implicitly, so we'll never see those, and `5XX` responses should presumably lead to a `5XX` response of our own; life is more interesting if we recieve a `4XX` response, but, again, those amount to errors that we want to report to the user, and the calling code shouldn't need to do anything more than propagate the error out to our interface.)
- The caller of the `_post()` (et al.) wrapper(s) wants the JSON payload, so (a) there's not much point in returning the `Response` object, and (b) if there's any problem unpacking the payload it would be better handled in the wrapper function.
- As for the contents of the payload, once it proves to be the expected JSON object, dealing with that I'm OK with leaving up to the caller.

Reply:
I believe that this conversation is now resolved except for the question of where/when to extract the JSON payload from the request.

I still think it would be better to do that at as low a level as feasible, to centralize (and DRY) the error handling.  However, two of the four wrappers (`_put()` and `_delete()`) currently have no callers, and two of the six calls to `_post()` (from functions that I don't think we want to keep -- `logout()` and `revoke_access_token()`) actually want the status, not the JSON.  So, until the dust settles a bit more, I guess we'll have to defer this.